✨ feat: support Liveblocks sync

CloudPassenger opened the pull request:
#### 💻 变更类型 | Change Type

<!-- For change type, change [ ] to [x]. -->

- [x] ✨ feat
- [ ] 🐛 fix
- [ ] ♻️ refactor
- [ ] 💄 style
- [ ] 🔨 chore
- [ ] ⚡️ perf
- [ ] 📝 docs

#### 🔀 变更说明 | Description of Change

This pull request introduces a new feature that improves data synchronisation in LobeChat by integrating the real-time online service provided by [Liveblocks](https://liveblocks.io/).

While the existing WebRTC solution provides a highly encrypted environment, it requires devices to be online at the same time and fails to synchronise data when P2P connections are unstable. By using servers hosted by Liveblocks, we can significantly improve the performance and reliability of data synchronisation.

During deployment, setting the `LIVEBLOCKS_SECRET_KEY` allows the use of the Liveblocks instance configured on the server. Users can also use their public key to connect to their own storage instance, providing a balance between convenience and security.

If login authentication is enabled in LobeChat, the user identifier is used as the sync room name by default, seamlessly integrating user authentication and data synchronisation.

Currently, the implementation has passed testing and can achieve bi-directional synchronisation between WebRTC P2P and the Liveblocks server. A demo video is attached:

https://github.com/lobehub/lobe-chat/assets/30863298/4561baa6-f3d8-4d17-9dff-aad8a7c453d0


#### 📝 补充信息 | Additional Information

This is a demo version that only integrates the new sync features. In a later update I will update more UI components like the `<SyncStatusInspector />` to make Liveblocks synchronisation compatible.

CloudPassenger committed:
✨ feat: support Liveblocks sync

arvinxx commented:
Amazing Work! plz rebase main branch~ 😅

arvinxx commented on src/app/(main)/settings/sync/features/Liveblocks/generateRandomRoomName.ts:
I think it's a common method , move it to the uilts folder?

arvinxx commented on src/config/app.ts:
```diff
@@ -50,6 +52,9 @@ export const getAppConfig = () => {
 
       PLUGIN_SETTINGS: process.env.PLUGIN_SETTINGS,
+
+      LIVEBLOCKS_SECRET_KEY: process.env.LIVEBLOCKS_SECRET_KEY || '',
```
if it's optional, no need the `|| ''`

arvinxx commented on src/config/featureFlags/index.ts:
```diff
@@ -22,6 +23,7 @@ export const FeatureFlagsSchema = z.object({
 export type IFeatureFlags = z.infer<typeof FeatureFlagsSchema>;
 
 export const DEFAULT_FEATURE_FLAGS: IFeatureFlags = {
+  liveblocks_sync: true,
```
I think it can be false by default, and enable it by add `+liveblocks_sync`

arvinxx commented on src/database/client/core/sync.ts:
```diff
@@ -48,108 +60,182 @@ class DataSync {
-    // ====== 2. init webrtc provider ====== //
-    this.logger(`[WebRTC] init provider... room: ${channel.name}`);
-    const { WebrtcProvider } = await import('y-webrtc');
+    if (channel.webrtc.enabled) {
```
I think there can be two methods: `connectWithWebRTC` and `connectWithLiveBlocks` ,

move the implement into these method and make the `connect` as entry.

arvinxx commented on src/features/SyncStatusInspector/DisableSync.tsx:
```diff
@@ -8,22 +8,24 @@ import { useTranslation } from 'react-i18next';
   const { t } = useTranslation('common');
   const [haveConfig, setSettings] = useUserStore((s) => [
-    !!syncSettingsSelectors.webrtcConfig(s).channelName,
+    !!channelSyncConfig(method)(s),
```
make `channelSyncConfig` as a selector

CloudPassenger committed:
♻️ refactor: liveblocks sync core

CloudPassenger committed:
🐛 fix: selectors for key vaults

ZeroDeng01 commented:
> Amazing Work! plz rebase main branch~ 😅

非常需要一个稳定的同步功能，希望早日上线👍

CloudPassenger commented:
Sorry, I've been very busy these days so my progress has slowed down, I'll try to keep on top of this pull request and make changes, I'm really sorry. :(

arvinxx commented:
> Sorry, I've been very busy these days so my progress has slowed down, I'll try to keep on top of this pull request and make changes, I'm really sorry. :(

Take you time, it not hurry for it~

CloudPassenger committed:
🐛 fix: userId missing when logged in

CloudPassenger committed:
♻️ refactor: adapt to latest lobechat db

CloudPassenger committed:
♻️ refactor(sync): Seperating liveblocks & webrtc

Introduce liveblocksDataSync and webrtcDataSync for real-time data synchronization. The changes were made to the index.ts, sync.ts, sync.test.ts, and action.ts files to implement the new features. The new classes extend the DataSync class and provide methods for starting data synchronization, initializing YDoc, observing changes in YMap, loading data from the database to Yjs, and managing awareness states. The StartDataSyncParams interface was updated to include the channel configuration for liveblocks and webrtc, and the CommonSyncParams interface was added to include common parameters for both types of synchronization. The SyncService class was also updated to enable and disable data synchronization based on the provided configuration.

CloudPassenger committed:
💄 style: support multiple sync channel tag

CloudPassenger committed:
🐛 fix: show sync settings separately

CloudPassenger committed:
🐛 fix: wrong status of sync service

CloudPassenger committed:
✨ feat: Liveblocks storage mode

CloudPassenger committed:
🐛 fix: use Hook mode to sync to Liveblocks

CloudPassenger committed:
♻️ refactor: update base model to adapt to Liveblocks

CloudPassenger committed:
🐛 fix: restore debug message

CloudPassenger committed:
🐛 fix: update Liveblocks dependencies

CloudPassenger committed:
Merge branch 'main' into feat/sync-liveblocks

CloudPassenger committed:
✅ test: fix test for Liveblocks sync

CloudPassenger committed:
🐛 fix: load flags environment correctly

CloudPassenger committed:
📝 docs: Liveblocks sync

CloudPassenger commented:
Hi,

I've made a large update to the Liveblocks sync feature.

- Use Liveblocks Storage instead of Liveblocks YJS to avoid additional data and space usage.
- At the same time, fix "Too large message error" (https://github.com/liveblocks/liveblocks/issues/1428 )
- Rewrite client-side sync logic to better handle local and remote data updates
- Updated self-host documentation for Liveblocks sync

arvinxx commented:
V2.0 会准备全面移除基于 dexie 的 db 存储逻辑，因此基于 YJS 的同步功能也会下线了。未来 client db 版本会准备加上 Webdav 的轻量同步功能。而更多潜在的同步能力未来估计会基于 pglite 和 postgres 的 CRDT 来探索，感谢你的贡献！
